src/withApollo.tsx

```diff
@@ -22,66 +20,85 @@ export default function withApollo<TCache = any>(
   client: InitApolloClient<TCache>,
   options: WithApolloOptions = {}
 ) {
-  type ApolloProps = WithApolloProps<TCache> & AppProps;
+  type ApolloProps = Partial<WithApolloProps<TCache>>;
 
-  if (!options.getDataFromTree) {
-    options.getDataFromTree = 'always';
+  if (options.getDataFromTree !== 'never') {
+    options.getDataFromTree = 'ssr';
   }
 
-  return (App: typeof NextApp) => {
-    return class WithApollo extends React.Component<ApolloProps> {
-      public static displayName = `WithApollo(${getDisplayName(App)})`;
+  return (
+    Page: NextPage<any> | typeof App,
+    pageOptions: WithApolloOptions = {}
+  ) => {
+    const getInitialProps = Page.getInitialProps;
+    const ssr =
+      pageOptions.getDataFromTree === 'never'
+        ? false
+        : options.getDataFromTree === 'ssr';
+    const render = pageOptions.render || options.render;
+
+    function WithApollo({ apollo, apolloState, ...props }: ApolloProps) {
+      const apolloClient =
+        apollo ||
+        initApollo<TCache>(client, { initialState: apolloState?.data });
+
+      if (render) {
+        return render({
+          Page: Page as NextPage<any>,
+          props: { ...props, apollo: apolloClient }
+        });
+      }
 
-      public static propTypes = {
-        apolloState: PropTypes.object,
-        apollo: PropTypes.object
-      };
+      return <Page {...props} apollo={apolloClient} />;
+    }
 
-      public static getInitialProps = async (appCtx: ApolloContext) => {
-        const { AppTree, ctx } = appCtx;
+    WithApollo.displayName = `WithApollo(${getDisplayName(Page)})`;
+
+    if (getInitialProps || ssr) {
+      WithApollo.getInitialProps = async (pageCtx: ApolloContext) => {
+        const ctx = 'Component' in pageCtx ? pageCtx.ctx : pageCtx;
```

Comment:
is it worth it in terms of readability to make this check a separate variable?
like `const isNextApp = 'Component' in pageCtx';` then also use it in L82?

Reply:
TypeScript complains about `pageCtx`, because it can be one of two types, and `'Component' in pageCtx` helps TypeScript knowing which one. Meaning TypeScript is not smart enough

Reply:
Ah I see, thank you for the clarification